bee-protocol/src/worker/propagator.rs

```diff
@@ -48,50 +49,66 @@ async fn propagate<B: StorageBackend>(
-            let new_ytrsi = max(parent1_ytrsi.unwrap(), parent2_ytrsi.unwrap());
-
-            let index = tangle
+            // Note: There are two types of solidity carriers:
+            // * solid messages (with available history)
+            // * solid entry points / sep (with verified history)
+            // Because we know both parents are solid, we also know that they have set OTRSI/YTRSI values, hence
+            // we can simply unwrap. We also try to minimise unnecessary Tangle API calls if - for example - the
+            // parent in question turns out to be a SEP.
+
+            // Determine OTRSI/YTRSI of parent1
+            let (parent1_otrsi, parent1_ytrsi) = match tangle.get_solid_entry_point_index(&parent1) {
+                Some(parent1_sepi) => (IndexId(parent1_sepi, parent1), IndexId(parent1_sepi, parent1)),
+                None => match tangle.get_metadata(&parent1).await {
+                    Some(parent1_md) => (parent1_md.otrsi().unwrap(), parent1_md.ytrsi().unwrap()),
+                    None => continue,
+                },
+            };
+
+            // Determine OTRSI/YTRSI of parent2
+            let (parent2_otrsi, parent2_ytrsi) = match tangle.get_solid_entry_point_index(&parent2) {
+                Some(parent2_sepi) => (IndexId(parent2_sepi, parent2), IndexId(parent2_sepi, parent2)),
+                None => match tangle.get_metadata(&parent2).await {
+                    Some(parent2_md) => (parent2_md.otrsi().unwrap(), parent2_md.ytrsi().unwrap()),
+                    None => continue,
+                },
+            };
+
+            // Derive child OTRSI/YTRSI from its parents.
+            // Note: The child inherits oldest (lowest) otrsi and the newest (highest) ytrsi from its parents.
+            let child_otrsi = min(parent1_otrsi, parent2_otrsi);
+            let child_ytrsi = max(parent1_ytrsi, parent2_ytrsi);
+
+            let ms_index_maybe = tangle
                 .update_metadata(&message_id, |metadata| {
-                    // OTRSI/YTRSI values need to be set before the solid flag, to ensure that the
-                    // MilestoneConeUpdater is aware of all values.
-                    metadata.set_otrsi(new_otrsi);
-                    metadata.set_ytrsi(new_ytrsi);
-                    metadata.solidify();
+                    // The child inherits the solid property from its parents.
+                    metadata.set_solid();
 
                     if metadata.flags().is_milestone() {
                         Some(metadata.milestone_index())
                     } else {
+                        metadata.set_otrsi(child_otrsi);
+                        metadata.set_ytrsi(child_ytrsi);
                         None
                     }
                 })
                 .await
                 .expect("Failed to fetch metadata.");
 
+            // Try to propagate as far as possible into the future.
```

Comment:
Could it be that the propagator and the confirmation worker trying to update the same future cone? Could it be that the propagator is faster? and we traverse the future cone another time in the confirmation worker?

Reply:
yea, both get step on each other's toes, so race conditions are very likely to happen. The propagator and the index updater **must not** run at the same time as they change the same metadata. There has to be a differentiation between a normal message and a milestone message:

Normal Message => run the propagator on that message as usual
Milestone Message => run the updater first, and *then* the propagator (but not before or during)

Reply:
We will add this later in another PR. Do you agree @thibault-martinez ?